src/txorphanage.h

```diff
@@ -21,40 +21,46 @@ extern RecursiveMutex g_cs_orphans;
 class TxOrphanage {
 public:
     /** Add a new orphan transaction */
-    bool AddTx(const CTransactionRef& tx, NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
+    bool AddTx(const CTransactionRef& tx, NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
     /** Check if we already have an orphan transaction (by txid or wtxid) */
-    bool HaveTx(const GenTxid& gtxid) const LOCKS_EXCLUDED(::g_cs_orphans);
-
-    /** Get an orphan transaction and its originating peer
-     * (Transaction ref will be nullptr if not found)
+    bool HaveTx(const GenTxid& gtxid) const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
+
+    /** Extract a transaction from a peer's work set
+     *  Returns nullptr and sets more to false if there are no transactions
+     *  to work on. Otherwise returns the transaction reference, removes
+     *  the transaction from the work set, and populates its arguments with
+     *  the originating peer, and whether there are more orphans for this peer
+     *  to work on after this tx.
      */
-    std::pair<CTransactionRef, NodeId> GetTx(const uint256& txid) const EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
+    CTransactionRef GetTxToReconsider(NodeId peer, NodeId& originator, bool& more) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
```

Comment:
```suggestion
    std::optional<TxToReconsider> GetTxToReconsider(NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
```
With `TxToReconsider` being:
```c++
struct TxToReconsider {
    const CTransactionRef tx;
    const NodeId originator;
    const bool more;
};
```

That would seem a little cleaner  to me because `originator` and `more` only have meaning if the returned transaction is not `nullptr`, so bundling the three would make sense imo.

Reply:
There was some other followup ideas in the previous PR, particularly https://github.com/bitcoin/bitcoin/pull/21527#discussion_r601221165 -- the idea there is that if Alice sends orphan tx Y which has a missing parent X, instead of immediately processing Y when X arrives (which might have arrived from Bob), just put it in Alice's workset, and delay processing of the orphan until it's Alice's turn. Then you'd have `peer == originator` and not need it to be returned as a result. At least last time I looked at this (the 202105-whohandlesorphans branch), it ended up simplifying the function to just `CTransactionRef GetTxToReconsider(NodeId peer)` (using `nullptr` as the "there's nothing to consider" indicator).

Reply:
Opened this as a draft at #26551